// neru-alpha/dist/esm/providers/state/state.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { StateCommand } from "./stateCommand.js";
import { StateOperations } from "./stateOperations.js";
import { RequestMethods } from "../../request/requestMethods.js";
/**
 * Represents a state object that can be used to store and retrieve data.
 * @class
 */
var State = /** @class */ (function () {
    /**
     * Creates a new State.
     * @constructor
     * @param {ISession} session - The session for the state object.
     * @param {string} namespace - The namespace for the state object.
     */
    function State(session, namespace) {
        if (namespace === void 0) { namespace = null; }
        /**
         * The provider for the state object.
         */
        this.provider = "client-persistence-api";
        this.bridge = session.bridge;
        this.url = session.config.getExecutionUrl(this.provider);
        if (namespace == null) {
            this.namespace = "state:".concat(session.id);
        }
        else {
            this.namespace = namespace;
        }
        this.session = session;
        this.commandService = session.commandService;
    }
    /**
     * Creates a new state command.
     * @param {string} op - The operation for the command.
     * @param {string} key - The key for the command.
     * @param {string[]} args - The arguments for the command.
     * @returns {StateCommand} The new state command.
     */
    State.prototype.createCommand = function (op, key, args) {
        return new StateCommand(op, this.namespace, key, args);
    };
    /**
     * Executes a state command.
     * @param {IStateCommand} command - The state command to execute.
     * @returns {Promise<T>} The result of the command.
     */
    State.prototype.executeCommand = function (command) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.commandService.executeCommand(this.url, RequestMethods.POST, command, this.session.constructRequestHeaders())];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Sets the value for a key in the state object.
     * @param {string} key - The key to set.
     * @param {T} value - The value to set for the key.
     * @returns {Promise<string>} A promise that resolves to "OK" if the operation was successful.
     */
    State.prototype.set = function (key, value) {
        return __awaiter(this, void 0, void 0, function () {
            var payload, command;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        payload = [];
                        payload.push(this.bridge.jsonStringify(value));
                        command = this.createCommand(StateOperations.SET, key, payload);
                        return [4 /*yield*/, this.executeCommand(command)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Retrieves the value for a key in the state object.
     * @param {string} key - The key to retrieve.
     * @returns {Promise<T>} A promise that resolves to the value for the key.
     */
    State.prototype.get = function (key) {
        return __awaiter(this, void 0, void 0, function () {
            var payload, command, result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        payload = [];
                        command = this.createCommand(StateOperations.GET, key, payload);
                        return [4 /*yield*/, this.executeCommand(command)];
                    case 1:
                        result = _a.sent();
                        if (result != null && result != "") {
                            return [2 /*return*/, this.bridge.jsonParse(result)];
                        }
                        return [2 /*return*/, null];
                }
            });
        });
    };
    /**
     * Deletes a key from the state object.
     * @param {string} key - The key to delete.
     * @returns {Promise<string>} A promise that resolves to "1" if the key was deleted, or "0" if the key did not exist.
     */
    State.prototype.delete = function (key) {
        return __awaiter(this, void 0, void 0, function () {
            var payload, command;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        payload = [];
                        command = this.createCommand(StateOperations.DEL, key, payload);
                        return [4 /*yield*/, this.executeCommand(command)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Deletes a key from a hash table in the state object.
     * @param {string} htable - The name of the hash table.
     * @param {string} key - The key to delete from the hash table.
     * @returns {Promise<string>} A promise that resolves to "1" if the key was deleted, or "0" if the key did not exist.
     */
    State.prototype.hdel = function (htable, keys) {
        return __awaiter(this, void 0, void 0, function () {
            var payload, command;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        payload = keys;
                        command = this.createCommand(StateOperations.HDEL, htable, payload);
                        return [4 /*yield*/, this.executeCommand(command)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Determines whether a key exists in a hash table in the state object.
     * @param {string} htable - The name of the hash table.
     * @param {string} key - The key to check for in the hash table.
     * @returns {Promise<string>} A promise that resolves to "1" if the key exists, or "0" if the key does not exist.
     */
    State.prototype.hexists = function (htable, key) {
        return __awaiter(this, void 0, void 0, function () {
            var payload, command;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        payload = [key];
                        command = this.createCommand(StateOperations.HEXISTS, htable, payload);
                        return [4 /*yield*/, this.executeCommand(command)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Retrieves all keys and values from a hash table in the state object.
     * @param {string} htable - The name of the hash table.
     * @returns {Promise<Record<string, string>>} A promise that resolves to an object containing the keys and values from the hash table.
     */
    State.prototype.hgetall = function (htable) {
        return __awaiter(this, void 0, void 0, function () {
            var payload, command, response, result, i;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        payload = [];
                        command = this.createCommand(StateOperations.HGETALL, htable, payload);
                        return [4 /*yield*/, this.executeCommand(command)];
                    case 1:
                        response = _a.sent();
                        result = {};
                        for (i = 0; i < response.length; i += 2) {
                            result[response[i]] = response[i + 1];
                        }
                        return [2 /*return*/, result];
                }
            });
        });
    };
    /**
     * Retrieves multiple values from a hash table in the state object.
     * @param {string} htable - The name of the hash table.
     * @param {string[]} keys - The keys to retrieve from the hash table.
     * @returns {Promise<string[]>} A promise that resolves to an array of values for the specified keys.
     */
    State.prototype.hmget = function (htable, keys) {
        return __awaiter(this, void 0, void 0, function () {
            var command, response, result, i;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        command = this.createCommand(StateOperations.HMGET, htable, keys);
                        return [4 /*yield*/, this.executeCommand(command)];
                    case 1:
                        response = _a.sent();
                        result = [];
                        for (i = 0; i < response.length; i++) {
                            result.push(response[i]);
                        }
                        return [2 /*return*/, result];
                }
            });
        });
    };
    /**
     * Retrieves all values from a hash table in the state object.
     * @param {string} htable - The name of the hash table.
     * @returns {Promise<string[]>} A promise that resolves to an array of values from the hash table.
     */
    State.prototype.hvals = function (htable) {
        return __awaiter(this, void 0, void 0, function () {
            var payload, command, response, result, i;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        payload = [];
                        command = this.createCommand(StateOperations.HVALS, htable, payload);
                        return [4 /*yield*/, this.executeCommand(command)];
                    case 1:
                        response = _a.sent();
                        result = [];
                        for (i = 0; i < response.length; i++) {
                            result.push(response[i]);
                        }
                        return [2 /*return*/, result];
                }
            });
        });
    };
    /**
     * Retrieves the value for a key in a hash table in the state object.
     * @param {string} htable - The name of the hash table.
     * @param {string} key - The key to retrieve from the hash table.
     * @returns {Promise<string>} A promise that resolves to the value for the key in the hash table.
     */
    State.prototype.hget = function (htable, key) {
        return __awaiter(this, void 0, void 0, function () {
            var payload, command, result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        payload = [key];
                        command = this.createCommand(StateOperations.HGET, htable, payload);
                        return [4 /*yield*/, this.executeCommand(command)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, result];
                }
            });
        });
    };
    /**
     * Sets multiple key-value pairs in a hash table in the state object.
     * @param {string} htable - The name of the hash table.
     * @param {Record<string, string>} keyValuePairs - An object containing the key-value pairs to set in the hash table.
     * @returns {Promise<string>} A promise that resolves to the number of fields that were added.
     */
    State.prototype.hset = function (htable, keyValuePairs) {
        return __awaiter(this, void 0, void 0, function () {
            var payload, keys, i, command;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        payload = [];
                        keys = this.bridge.getObjectKeys(keyValuePairs);
                        for (i = 0; i < keys.length; i++) {
                            payload.push(keys[i]);
                            payload.push(keyValuePairs[keys[i]]);
                        }
                        command = this.createCommand(StateOperations.HSET, htable, payload);
                        return [4 /*yield*/, this.executeCommand(command)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Increments the value for a key in a hash table in the state object.
     * @param {string} htable The name of the hash table.
     * @param {string} key The key to increment.
     * @param {number} value The amount to increment the key by.
     * @returns {Promise<string>} A promise that resolves to the new value for the key.
     */
    State.prototype.hincrby = function (htable, key, value) {
        return __awaiter(this, void 0, void 0, function () {
            var payload, command;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        payload = [key, this.bridge.jsonStringify(value)];
                        command = this.createCommand(StateOperations.HINCRBY, htable, payload);
                        return [4 /*yield*/, this.executeCommand(command)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Retrieves the number of key-value pairs in a hash table in the state object.
     * @param {string} htable - The name of the hash table.
     * @returns {Promise<string>} A promise that resolves to the number of key-value pairs in the hash table.
     */
    State.prototype.hlen = function (htable) {
        return __awaiter(this, void 0, void 0, function () {
            var payload, command;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        payload = [];
                        command = this.createCommand(StateOperations.HLEN, htable, payload);
                        return [4 /*yield*/, this.executeCommand(command)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    State.prototype.hscan = function (htable, cursor, pattern, count) {
        return __awaiter(this, void 0, void 0, function () {
            var payload, command;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        payload = [cursor, "MATCH", pattern, "COUNT", this.bridge.jsonStringify(count)];
                        command = this.createCommand(StateOperations.HSCAN, htable, payload);
                        return [4 /*yield*/, this.executeCommand(command)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Adds a value to the end of a list in the state object.
     * @param {string} list - The name of the list.
     * @param {T} value - The value to add to the list.
     * @returns {Promise<string>} A promise that resolves to the new length of the list.
     */
    State.prototype.rpush = function (list, value) {
        return __awaiter(this, void 0, void 0, function () {
            var payload, command;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        payload = [this.bridge.jsonStringify(value)];
                        command = this.createCommand(StateOperations.RPUSH, list, payload);
                        return [4 /*yield*/, this.executeCommand(command)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Removes and retrieves one or more values from the end of a list in the state object.
     * @param {string} list - The name of the list.
     * @param {number} count - The number of values to remove and retrieve from the end of the list.
     * @returns {Promise<T[]>} A promise that resolves to an array of values removed from the end of the list.
     */
    State.prototype.rpop = function (list, count) {
        if (count === void 0) { count = 1; }
        return __awaiter(this, void 0, void 0, function () {
            var args, command, response, result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        args = [this.bridge.jsonStringify(count)];
                        command = this.createCommand(StateOperations.RPOP, list, args);
                        return [4 /*yield*/, this.executeCommand(command)];
                    case 1:
                        response = _a.sent();
                        result = this.parseResponse(response);
                        return [2 /*return*/, result];
                }
            });
        });
    };
    /**
     * Adds a value to the beginning of a list in the state object.
     * @param {string} list - The name of the list.
     * @param {T} value - The value to add to the list.
     * @returns {Promise<string>} A promise that resolves to the new length of the list.
     */
    State.prototype.lpush = function (list, value) {
        return __awaiter(this, void 0, void 0, function () {
            var payload, command;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        payload = [this.bridge.jsonStringify(value)];
                        command = this.createCommand(StateOperations.LPUSH, list, payload);
                        return [4 /*yield*/, this.executeCommand(command)];
                    case 1: return [2 /*return*/, _a.sent()]; // returns the array length  e.g. "1"
                }
            });
        });
    };
    /**
     * Removes and retrieves one or more values from the beginning of a list in the state object.
     * @param {string} list - The name of the list.
     * @param {number} count - The number of values to remove and retrieve from the beginning of the list. Default is 1.
     * @returns {Promise<T[]>} A promise that resolves to an array of values removed from the beginning of the list.
     */
    State.prototype.lpop = function (list, count) {
        if (count === void 0) { count = 1; }
        return __awaiter(this, void 0, void 0, function () {
            var args, command, response, result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        args = [this.bridge.jsonStringify(count)];
                        command = this.createCommand(StateOperations.LPOP, list, args);
                        return [4 /*yield*/, this.executeCommand(command)];
                    case 1:
                        response = _a.sent();
                        result = this.parseResponse(response);
                        return [2 /*return*/, result];
                }
            });
        });
    };
    /**
     * Removes one or more occurrences of a value from a list in the state object.
     * @param {string} list - The name of the list.
     * @param {T} value - The value to remove from the list.
     * @param {number} count - The number of occurrences to remove. A positive number removes the first `count` occurrences, a negative number removes the last `count` occurrences, and a value of 0 removes all occurrences.
     * @returns {Promise<string>} A promise that resolves to the number of occurrences removed from the list.
     */
    State.prototype.lrem = function (list, value, count) {
        if (count === void 0) { count = 0; }
        return __awaiter(this, void 0, void 0, function () {
            var args, command;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        args = [this.bridge.jsonStringify(count), this.bridge.jsonStringify(value)];
                        command = this.createCommand(StateOperations.LREM, list, args);
                        return [4 /*yield*/, this.executeCommand(command)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Retrieves the length of a list in the state object.
     * @param {string} list - The name of the list.
     * @returns {Promise<string>} A promise that resolves to the length of the list.
     */
    State.prototype.llen = function (list) {
        return __awaiter(this, void 0, void 0, function () {
            var payload, command;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        payload = [];
                        command = this.createCommand(StateOperations.LLEN, list, payload);
                        return [4 /*yield*/, this.executeCommand(command)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Retrieves a range of values from a list in the state object.
     * @param {string} list - The name of the list.
     * @param {number} startPos - The starting position of the range.
     * @param {number} endPos - The ending position of the range.
     * @returns {Promise<T[]>} A promise that resolves to an array of values from the specified range in the list.
     */
    State.prototype.lrange = function (list, startPos, endPos) {
        if (startPos === void 0) { startPos = 0; }
        if (endPos === void 0) { endPos = -1; }
        return __awaiter(this, void 0, void 0, function () {
            var args, command, response, result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        args = [this.bridge.jsonStringify(startPos), this.bridge.jsonStringify(endPos)];
                        command = this.createCommand(StateOperations.LRANGE, list, args);
                        return [4 /*yield*/, this.executeCommand(command)];
                    case 1:
                        response = _a.sent();
                        result = this.parseResponse(response);
                        return [2 /*return*/, result];
                }
            });
        });
    };
    /**
     * Trims a list in the state object to contain only the specified range of values.
     * @param {string} list - The name of the list.
     * @param {number} startPos - The starting position of the range.
     * @param {number} endPos - The ending position of the range.
     * @returns {Promise<string>} A promise that resolves to "OK" if the operation was successful.
     */
    State.prototype.ltrim = function (list, startPos, endPos) {
        return __awaiter(this, void 0, void 0, function () {
            var args, command, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        args = [this.bridge.jsonStringify(startPos), this.bridge.jsonStringify(endPos)];
                        command = this.createCommand(StateOperations.LTRIM, list, args);
                        return [4 /*yield*/, this.executeCommand(command)];
                    case 1:
                        response = _a.sent();
                        return [2 /*return*/, response];
                }
            });
        });
    };
    /**
     * Inserts a value into a list in the state object either before or after a pivot value.
     * @param {string} list - The name of the list.
     * @param {boolean} before - If `true`, the value is inserted before the pivot value. If `false`, the value is inserted after the pivot value.
     * @param {T} pivot - The pivot value.
     * @param {T} value - The value to insert into the list.
     * @returns {Promise<string>} A promise that resolves to the new length of the list if the operation was successful.
     */
    State.prototype.linsert = function (list, before, pivot, value) {
        return __awaiter(this, void 0, void 0, function () {
            var direction, args, command, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        direction = "AFTER";
                        if (before == true) {
                            direction = "BEFORE";
                        }
                        args = [direction, this.bridge.jsonStringify(pivot), this.bridge.jsonStringify(value)];
                        command = this.createCommand(StateOperations.LINSERT, list, args);
                        return [4 /*yield*/, this.executeCommand(command)];
                    case 1:
                        response = _a.sent();
                        return [2 /*return*/, response];
                }
            });
        });
    };
    /**
     * Retrieves the value at a specific position in a list in the state object.
     * @param {string} list - The name of the list.
     * @param {number} position - The position of the value to retrieve.
     * @returns {Promise<T>} A promise that resolves to the value at the specified position in the list.
     */
    State.prototype.lindex = function (list, position) {
        return __awaiter(this, void 0, void 0, function () {
            var args, command, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        args = [this.bridge.jsonStringify(position)];
                        command = this.createCommand(StateOperations.LINDEX, list, args);
                        return [4 /*yield*/, this.executeCommand(command)];
                    case 1:
                        response = _a.sent();
                        return [2 /*return*/, this.bridge.jsonParse(response)];
                }
            });
        });
    };
    /**
     * Sets the value at a specific position in a list in the state object.
     * @param {string} list - The name of the list.
     * @param {number} position - The position of the value to set.
     * @param {T} value - The value to set at the specified position in the list.
     * @returns {Promise<string>} A promise that resolves to "OK" if the operation was successful.
     */
    State.prototype.lset = function (list, position, value) {
        return __awaiter(this, void 0, void 0, function () {
            var args, command;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        args = [this.bridge.jsonStringify(position), this.bridge.jsonStringify(value)];
                        command = this.createCommand(StateOperations.LSET, list, args);
                        return [4 /*yield*/, this.executeCommand(command)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Increments the value of a key in the state object by a specified amount.
     * @param {string} key - The key to increment.
     * @param {number} value - The amount to increment the key by.
     * @returns {Promise<string>} A promise that resolves to the new value of the key.
     */
    State.prototype.incrby = function (key, value) {
        return __awaiter(this, void 0, void 0, function () {
            var args, command, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        args = [this.bridge.jsonStringify(value)];
                        command = this.createCommand(StateOperations.INCRBY, key, args);
                        return [4 /*yield*/, this.executeCommand(command)];
                    case 1:
                        response = _a.sent();
                        return [2 /*return*/, this.bridge.jsonParse(response)];
                }
            });
        });
    };
    /**
     * Decrements the value of a key in the state object by a specified amount.
     * @param {string} key - The key to decrement.
     * @param {number} value - The amount to decrement the key by.
     * @returns {Promise<string>} A promise that resolves to the new value of the key.
     */
    State.prototype.decrby = function (key, value) {
        return __awaiter(this, void 0, void 0, function () {
            var args, command, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        args = [this.bridge.jsonStringify(value)];
                        command = this.createCommand(StateOperations.DECRBY, key, args);
                        return [4 /*yield*/, this.executeCommand(command)];
                    case 1:
                        response = _a.sent();
                        return [2 /*return*/, this.bridge.jsonParse(response)];
                }
            });
        });
    };
    /**
     * Sets an expiration time for a key in the state object.
     * @param {string} key - The key to set the expiration time for.
     * @param {number} seconds - The number of seconds after which the key will expire.
     * @param {ExpireOptions} [option] - An optional object specifying additional options for the operation.
     * @returns {Promise<string>} A promise that resolves to "1" if the operation was successful, or "0" if the key does not exist.
     */
    State.prototype.expire = function (key, seconds, option) {
        return __awaiter(this, void 0, void 0, function () {
            var args, command;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        args = [this.bridge.jsonStringify(seconds)];
                        if (option != null) {
                            args.push(option);
                        }
                        command = this.createCommand(StateOperations.EXPIRE, key, args);
                        return [4 /*yield*/, this.executeCommand(command)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    State.prototype.parseResponse = function (response) {
        var result = [];
        if (response != null) {
            for (var i = 0; i < response.length; i++) {
                result.push(this.bridge.jsonParse(response[i]));
            }
        }
        return result;
    };
    return State;
}());
export { State };
