// neru-alpha/dist/esm/providers/state/stateOperations.js
export var StateOperations;
(function (StateOperations) {
    StateOperations["SET"] = "SET";
    StateOperations["GET"] = "GET";
    StateOperations["DEL"] = "DEL";
    StateOperations["HDEL"] = "HDEL";
    StateOperations["HEXISTS"] = "HEXISTS";
    StateOperations["HGETALL"] = "HGETALL";
    StateOperations["HMGET"] = "HMGET";
    StateOperations["HVALS"] = "HVALS";
    StateOperations["HGET"] = "HGET";
    StateOperations["HSET"] = "HSET";
    StateOperations["RPUSH"] = "RPUSH";
    StateOperations["LPUSH"] = "LPUSH";
    StateOperations["LLEN"] = "LLEN";
    StateOperations["LRANGE"] = "LRANGE";
    StateOperations["LPOP"] = "LPOP";
    StateOperations["LTRIM"] = "LTRIM";
    StateOperations["LINSERT"] = "LINSERT";
    StateOperations["LINDEX"] = "LINDEX";
    StateOperations["LSET"] = "LSET";
    StateOperations["LREM"] = "LREM";
    StateOperations["INCRBY"] = "INCRBY";
    StateOperations["DECRBY"] = "DECRBY";
    StateOperations["RPOP"] = "RPOP";
    StateOperations["EXPIRE"] = "EXPIRE";
    StateOperations["HLEN"] = "HLEN";
    StateOperations["HINCRBY"] = "HINCRBY";
    StateOperations["HSCAN"] = "HSCAN";
})(StateOperations || (StateOperations = {}));
