// @vonage/vcr-sdk/dist/esm/providers/scheduler/scheduler.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { SCHEDULER_ACTION } from "./enums/schedulerAction.js";
import { StartAtPayload } from "./contracts/startAtPayload.js";
import { SchedulerPayload } from "./contracts/schedulerPayload.js";
import { ListAllPayload } from "./contracts/listAllPayload.js";
import { REQUEST_VERB } from "../../request/enums/requestVerb.js";
import { RequestParams } from "../../request/contracts/requestParams.js";
var Scheduler = /** @class */ (function () {
    function Scheduler(session) {
        this.apiVersion = "v2";
        this.provider = "vonage-scheduler";
        this.session = session;
        this.bridge = session.bridge;
    }
    Scheduler.prototype.startAt = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var payload, requestParams, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (params.id != null && this.bridge.testRegEx(params.id, "^[a-zA-Z0-9][a-zA-Z0-9-_]*$") != true) {
                            throw this.bridge.createSdkError("startAt: The input does not match the required pattern ^[a-zA-Z0-9][a-zA-Z0-9-_]*$. Please enter a string that starts with a letter or a digit, and contains only letters, digits, hyphens, and underscores.");
                        }
                        payload = new StartAtPayload();
                        payload.startAt = params.startAt;
                        payload.callback = this.session.wrapCallback(params.callback, []);
                        if (params.payload != null) {
                            payload.payload = params.payload;
                        }
                        if (params.interval != null) {
                            payload.interval = params.interval;
                        }
                        if (params.id != null) {
                            payload.id = params.id;
                        }
                        requestParams = new RequestParams();
                        requestParams.method = REQUEST_VERB.POST;
                        requestParams.data = payload;
                        requestParams.url = this.session.config.getExecutionUrl(this.provider, "".concat(this.apiVersion, "/").concat(SCHEDULER_ACTION.CREATE));
                        requestParams.headers = this.session.constructRequestHeaders();
                        return [4 /*yield*/, this.session.request(requestParams)];
                    case 1:
                        response = _a.sent();
                        return [2 /*return*/, response["id"]];
                }
            });
        });
    };
    Scheduler.prototype.list = function (size, cursor) {
        if (size === void 0) { size = 10; }
        return __awaiter(this, void 0, void 0, function () {
            var payload, requestParams;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        payload = new ListAllPayload(size, cursor);
                        requestParams = new RequestParams();
                        requestParams.method = REQUEST_VERB.POST;
                        requestParams.data = payload;
                        requestParams.url = this.session.config.getExecutionUrl(this.provider, "".concat(this.apiVersion, "/").concat(SCHEDULER_ACTION.LIST));
                        requestParams.headers = this.session.constructRequestHeaders();
                        return [4 /*yield*/, this.session.request(requestParams)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    Scheduler.prototype.get = function (scheduleId) {
        return __awaiter(this, void 0, void 0, function () {
            var payload, requestParams;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        payload = new SchedulerPayload(scheduleId);
                        requestParams = new RequestParams();
                        requestParams.method = REQUEST_VERB.POST;
                        requestParams.data = payload;
                        requestParams.url = this.session.config.getExecutionUrl(this.provider, "".concat(this.apiVersion, "/").concat(SCHEDULER_ACTION.GET));
                        requestParams.headers = this.session.constructRequestHeaders();
                        return [4 /*yield*/, this.session.request(requestParams)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    Scheduler.prototype.cancel = function (scheduleId) {
        return __awaiter(this, void 0, void 0, function () {
            var payload, requestParams;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        payload = new SchedulerPayload(scheduleId);
                        requestParams = new RequestParams();
                        requestParams.method = REQUEST_VERB.POST;
                        requestParams.data = payload;
                        requestParams.url = this.session.config.getExecutionUrl(this.provider, "".concat(this.apiVersion, "/").concat(SCHEDULER_ACTION.CANCEL));
                        requestParams.headers = this.session.constructRequestHeaders();
                        return [4 /*yield*/, this.session.request(requestParams)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/, payload.id];
                }
            });
        });
    };
    return Scheduler;
}());
export { Scheduler };
