// @vonage/vcr-sdk/dist/esm/providers/voice/voice.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { VOICE_ACTION } from "./enums/voiceAction.js";
import { CallAnswerCallBack } from "./contracts/callAnswerCallBack.js";
import { CallEventCallBackPayload } from "./contracts/CallEventCallBackPayload.js";
import { RequestParams } from "../../request/contracts/requestParams.js";
import { CreateVonageTokenParams } from "../../services/jwt/contracts/createVonageTokenParams.js";
import { Assets } from "../assets/assets.js";
import { REQUEST_VERB } from "../../request/enums/requestVerb.js";
var Voice = /** @class */ (function () {
    function Voice(session) {
        this.provider = "vonage-voice";
        this.session = session;
        this.bridge = session.bridge;
        this.assetsAPI = new Assets(session);
    }
    Voice.prototype.onCall = function (callback) {
        return __awaiter(this, void 0, void 0, function () {
            var requestParams;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        requestParams = new RequestParams();
                        requestParams.method = REQUEST_VERB.POST;
                        requestParams.data = new CallAnswerCallBack(this.session.wrapCallback(callback, []));
                        requestParams.headers = this.session.constructRequestHeaders();
                        requestParams.url = this.session.config.getExecutionUrl(this.provider, VOICE_ACTION.VAPI_SUBSCRIBE_INBOUND_CALL);
                        return [4 /*yield*/, this.session.request(requestParams)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/, requestParams.data.callback.id];
                }
            });
        });
    };
    Voice.prototype.onCallEvent = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var payload, requestParams;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        payload = new CallEventCallBackPayload();
                        payload.callback = this.session.wrapCallback(params.callback, []);
                        if (params.vapiID != null) {
                            payload.vapiId = params.vapiID;
                        }
                        else if (params.conversationID != null) {
                            payload.conversationId = params.conversationID;
                        }
                        requestParams = new RequestParams();
                        requestParams.method = REQUEST_VERB.POST;
                        requestParams.data = payload;
                        requestParams.headers = this.session.constructRequestHeaders();
                        requestParams.url = this.session.config.getExecutionUrl(this.provider, VOICE_ACTION.VAPI_SUBSCRIBE_EVENT);
                        return [4 /*yield*/, this.session.request(requestParams)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/, payload.callback.id];
                }
            });
        });
    };
    Voice.prototype.getCallRecording = function (recordingUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var tp, t, rp;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        tp = new CreateVonageTokenParams();
                        tp.exp = this.bridge.getSystemTime() + 60 * 60;
                        t = this.session.jwt.createVonageToken(tp);
                        rp = new RequestParams();
                        rp.method = REQUEST_VERB.GET;
                        rp.url = recordingUrl;
                        rp.headers = {};
                        rp.headers["Authorization"] = "Bearer ".concat(t);
                        rp.responseType = "stream" /* RESPONSE_TYPE.STREAM */;
                        return [4 /*yield*/, this.session.request(rp)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    Voice.prototype.uploadCallRecording = function (recordingUrl, assetsPath) {
        return __awaiter(this, void 0, void 0, function () {
            var stream, pathObject, data, fileNames;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getCallRecording(recordingUrl)];
                    case 1:
                        stream = _a.sent();
                        pathObject = this.bridge.parsePath(assetsPath);
                        data = [stream];
                        fileNames = [pathObject.base];
                        return [4 /*yield*/, this.assetsAPI.uploadData(data, pathObject.dir, fileNames)];
                    case 2: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    return Voice;
}());
export { Voice };
