// neru-alpha/dist/esm/services/config/config.js
var Config = /** @class */ (function () {
    function Config(bridge) {
        this.instanceId = "debug";
        this.debug = false;
        this.logsSubmission = true;
        this.namespace = "neru";
        this.bridge = bridge;
        if (this.bridge.getEnv("NAMESPACE") != null) {
            this.namespace = this.bridge.getEnv("NAMESPACE");
        }
        this.instanceServiceName = this.bridge.getEnv("INSTANCE_SERVICE_NAME");
        this.applicationId = this.bridge.getEnv("APPLICATION_ID");
        if (this.bridge.getEnv("INSTANCE_ID") != null) {
            this.instanceId = this.bridge.getEnv("INSTANCE_ID");
        }
        this.apiApplicationId = this.bridge.getEnv("API_APPLICATION_ID");
        this.apiAccountId = this.bridge.getEnv("API_ACCOUNT_ID");
        this.privateKey = this.bridge.getEnv("PRIVATE_KEY");
        this.appUrl = this.bridge.getEnv("VCR_INSTANCE_PUBLIC_URL");
        var debug = this.bridge.getEnv("DEBUG");
        if (debug == "true") {
            this.debug = true;
        }
        this.assetUrl = "http://openfaas.euw1.dev.nexmo.cloud/function/vonage-assets?get=";
    }
    Config.prototype.getExecutionUrl = function (func, pathname, queryParams) {
        if (pathname === void 0) { pathname = null; }
        var hostname = "".concat(func, ".").concat(this.namespace);
        if (this.debug) {
            return this.constructDebugUrl(hostname, pathname, queryParams);
        }
        return this.constructProdUrl(hostname, pathname, queryParams);
    };
    Config.prototype.constructDebugUrl = function (hostname, pathname, queryParams) {
        var url = "http://localhost:3001";
        if (pathname != null) {
            url += "/".concat(pathname);
        }
        url += "?func=".concat(hostname, "&async=false");
        if (queryParams != null) {
            url += "&";
            url += this.queryObjectToString(queryParams);
        }
        return url;
    };
    Config.prototype.constructProdUrl = function (hostname, pathname, queryParams) {
        var url = "http://".concat(hostname);
        if (pathname != null) {
            url += "/".concat(pathname);
        }
        if (queryParams != null) {
            url += "?";
            url += this.queryObjectToString(queryParams);
        }
        return url;
    };
    Config.prototype.queryObjectToString = function (queryObject) {
        var keys = this.bridge.getObjectKeys(queryObject);
        var queryString = "";
        for (var i = 0; i < keys.length; i++) {
            var key = keys[i]; // Original key
            if (queryObject[key] != null) {
                var encodedKey = this.bridge.encodeUriComponent(key); // Encoded key
                var value = this.bridge.encodeUriComponent(queryObject[key]);
                queryString += "".concat(encodedKey, "=").concat(value);
                if (i < keys.length - 1) {
                    queryString += "&";
                }
            }
        }
        return queryString;
    };
    return Config;
}());
export { Config };
