// neru-alpha/dist/esm/providers/meetings/meetings.js
import { VonageAPI } from "../vonageAPI/vonageAPI.js";
import { DeleteRoomPayload } from "./contracts/deleteRoomPayload.js";
import { UpdateRoomDetails } from "./contracts/updateRoomDetails.js";
import { UpdateRoomPayload } from "./contracts/updateRoomPayload.js";
import { RequestMethods } from "../../request/requestMethods.js";
var Meetings = /** @class */ (function () {
    function Meetings(session) {
        this.session = session;
        this.vonageAPI = new VonageAPI(this.session);
        this.baseUrl = "https://api-eu.vonage.com/beta/meetings";
    }
    Meetings.prototype.getRoom = function (roomId) {
        var url = "".concat(this.baseUrl, "/rooms/").concat(roomId);
        var method = RequestMethods.GET;
        return this.vonageAPI.invoke(url, method, null);
    };
    Meetings.prototype.getRooms = function (paginationUrl) {
        var url = "".concat(this.baseUrl, "/rooms");
        if (paginationUrl != null) {
            url = paginationUrl;
        }
        var method = RequestMethods.GET;
        return this.vonageAPI.invoke(url, method, null);
    };
    Meetings.prototype.createRoom = function (createRoomPayload) {
        var url = "".concat(this.baseUrl, "/rooms");
        var method = RequestMethods.POST;
        return this.vonageAPI.invoke(url, method, createRoomPayload);
    };
    Meetings.prototype.updateRoom = function (roomId, expiry, expireAfterUse) {
        var details = new UpdateRoomDetails(expiry, expireAfterUse);
        var payload = new UpdateRoomPayload(details);
        var url = "".concat(this.baseUrl, "/rooms/").concat(roomId);
        var method = RequestMethods.PATCH;
        return this.vonageAPI.invoke(url, method, payload);
    };
    Meetings.prototype.deleteRoom = function (roomId) {
        var url = "".concat(this.baseUrl, "/rooms/").concat(roomId);
        var method = RequestMethods.DEL;
        var payload = new DeleteRoomPayload();
        return this.vonageAPI.invoke(url, method, payload);
    };
    return Meetings;
}());
export { Meetings };
