// @vonage/vcr-sdk/dist/cjs/vcr.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.vcr = exports.VCR = void 0;
const vcrSession_js_1 = require("./session/vcrSession.js");
const jwt_js_1 = require("./services/jwt/jwt.js");
const bridge_js_1 = require("./bridge.js");
const config_js_1 = require("./services/config/config.js");
const state_js_1 = require("./providers/state/state.js");
const unsubscribeEventsPayload_js_1 = require("./providers/messages/contracts/unsubscribeEventsPayload.js");
const actionPayload_js_1 = require("./session/contracts/actionPayload.js");
const messageAction_js_1 = require("./providers/messages/enums/messageAction.js");
const requestParams_js_1 = require("./request/contracts/requestParams.js");
const command_js_1 = require("./session/contracts/command.js");
const requestVerb_js_1 = require("./request/enums/requestVerb.js");
class VCR {
    constructor() {
        this.bridge = new bridge_js_1.Bridge();
        this.config = new config_js_1.Config(this.bridge);
        this.jwt = new jwt_js_1.JWT(this.bridge, this.config);
    }
    unsubscribe(id, provider = "vonage-messaging") {
        return __awaiter(this, void 0, void 0, function* () {
            const p = new unsubscribeEventsPayload_js_1.UnsubscribeEventsPayload(id);
            const a = new actionPayload_js_1.ActionPayload(provider, messageAction_js_1.MESSAGE_ACTION.UNSUBSCRIBE_EVENTS, p);
            const s = this.getGlobalSession();
            const ch = s.constructCommandHeaders();
            const rp = new requestParams_js_1.RequestParams();
            rp.method = requestVerb_js_1.REQUEST_VERB.POST;
            rp.data = new command_js_1.Command(ch, a);
            rp.headers = s.constructRequestHeaders();
            rp.url = s.config.getExecutionUrl(provider);
            return yield s.request(rp);
        });
    }
    createVonageToken(params) {
        if (params == null) {
            throw this.bridge.createSdkError("createVonageToken: params is required");
        }
        if (params.exp == null) {
            throw this.bridge.createSdkError("createVonageToken: params.exp is required");
        }
        return this.jwt.createVonageToken(params);
    }
    /**
     * Create a new session
     * @param ttl time to live in seconds, defaults to 7 days
     * @returns
     */
    createSession(ttl = 7 * 24 * 60 * 60) {
        // validate ttl is a positive integer
        if (this.bridge.isInteger(ttl) == false || ttl < 0) {
            throw this.bridge.createSdkError("createSession: ttl must be a positive integer");
        }
        const id = this.bridge.uuid();
        const session = this.createSessionWithId(id);
        this.bridge.runBackgroundTask(session.emitSessionCreatedEvent(ttl));
        return session;
    }
    createSessionWithId(id) {
        return new vcrSession_js_1.VCRSession(this.bridge, this.config, this.jwt, id);
    }
    getSessionById(id) {
        if (id == null) {
            throw this.bridge.createSdkError("getSessionById: id is required");
        }
        return this.createSessionWithId(id);
    }
    getAppUrl() {
        return this.config.appUrl;
    }
    getSessionFromRequest(req) {
        if (req == null) {
            throw this.bridge.createSdkError("getSessionFromRequest: function requires request object to be provided");
        }
        if (req.headers == null) {
            throw this.bridge.createSdkError("getSessionFromRequest: invalid request object proivided");
        }
        const id = req.headers["x-neru-sessionid"];
        if (id == null) {
            throw this.bridge.createSdkError(`getSessionFromRequest: request does not contain \"x-neru-sessionid\" header`);
        }
        return this.getSessionById(id);
    }
    getGlobalSession() {
        const uuid = "00000000-0000-0000-0000-000000000000";
        return this.getSessionById(uuid);
    }
    /**
     * Returns the instance state that is shared across all sessions within the same instance.
     */
    getInstanceState() {
        const session = this.getGlobalSession();
        return new state_js_1.State(session, `application:${this.config.instanceId}`);
    }
    /**
     * Returns the account state that is shared across all applications within the same account.
     */
    getAccountState() {
        const session = this.getGlobalSession();
        return new state_js_1.State(session, "account");
    }
}
exports.VCR = VCR;
exports.vcr = new VCR();
// IMPORTANT: autogenerated code starts here, DO NOT DELETE OR CHANGE THIS COMMENT.
__exportStar(require("./bridge.js"), exports);
__exportStar(require("./IVCR.js"), exports);
__exportStar(require("./IBridge.js"), exports);
__exportStar(require("./webhookEvents/urlPayload.js"), exports);
__exportStar(require("./webhookEvents/baseEvent.js"), exports);
__exportStar(require("./webhookEvents/IUrlPayload.js"), exports);
__exportStar(require("./webhookEvents/IBaseEvent.js"), exports);
__exportStar(require("./session/vcrSession.js"), exports);
__exportStar(require("./session/ISession.js"), exports);
__exportStar(require("./errors/ISdkError.js"), exports);
__exportStar(require("./webhookEvents/enums/webhookEventType.js"), exports);
__exportStar(require("./services/jwt/jwt.js"), exports);
__exportStar(require("./services/jwt/IJwt.js"), exports);
__exportStar(require("./services/config/pathObject.js"), exports);
__exportStar(require("./services/config/config.js"), exports);
__exportStar(require("./services/config/IPathObject.js"), exports);
__exportStar(require("./services/config/IConfig.js"), exports);
__exportStar(require("./session/contracts/wrappedCallback.js"), exports);
__exportStar(require("./session/contracts/filter.js"), exports);
__exportStar(require("./session/contracts/command.js"), exports);
__exportStar(require("./session/contracts/actionPayload.js"), exports);
__exportStar(require("./session/contracts/IWrappedCallback.js"), exports);
__exportStar(require("./session/contracts/IPayloadWithCallback.js"), exports);
__exportStar(require("./session/contracts/IFilter.js"), exports);
__exportStar(require("./session/contracts/ICommand.js"), exports);
__exportStar(require("./session/contracts/IActionPayload.js"), exports);
__exportStar(require("./providers/vonageAPI/vonageAPIActions.js"), exports);
__exportStar(require("./providers/vonageAPI/vonageAPI.js"), exports);
__exportStar(require("./providers/vonageAPI/IVonageAPI.js"), exports);
__exportStar(require("./providers/voice/voice.js"), exports);
__exportStar(require("./providers/voice/IVoice.js"), exports);
__exportStar(require("./providers/state/state.js"), exports);
__exportStar(require("./providers/state/IState.js"), exports);
__exportStar(require("./providers/scheduler/scheduler.js"), exports);
__exportStar(require("./providers/scheduler/IScheduler.js"), exports);
__exportStar(require("./providers/queue/queue.js"), exports);
__exportStar(require("./providers/queue/IQueue.js"), exports);
__exportStar(require("./providers/messages/messages.js"), exports);
__exportStar(require("./providers/messages/IMessages.js"), exports);
__exportStar(require("./providers/logger/logger.js"), exports);
__exportStar(require("./providers/logger/ILogger.js"), exports);
__exportStar(require("./providers/events/eventFactory.js"), exports);
__exportStar(require("./providers/events/eventEmitter.js"), exports);
__exportStar(require("./providers/events/IEventFactory.js"), exports);
__exportStar(require("./providers/events/IEventEmitter.js"), exports);
__exportStar(require("./providers/assets/assets.js"), exports);
__exportStar(require("./providers/assets/IAssets.js"), exports);
__exportStar(require("./providers/conversation/conversation.js"), exports);
__exportStar(require("./providers/conversation/IConversation.js"), exports);
__exportStar(require("./request/enums/responseTypes.js"), exports);
__exportStar(require("./request/enums/requestVerb.js"), exports);
__exportStar(require("./request/enums/dataType.js"), exports);
__exportStar(require("./request/contracts/requestParams.js"), exports);
__exportStar(require("./request/contracts/request.js"), exports);
__exportStar(require("./request/contracts/formDataObject.js"), exports);
__exportStar(require("./request/contracts/IRequestParams.js"), exports);
__exportStar(require("./request/contracts/IRequest.js"), exports);
__exportStar(require("./request/contracts/IFormDataObject.js"), exports);
__exportStar(require("./services/jwt/enums/algorithm.js"), exports);
__exportStar(require("./services/jwt/contracts/vonageJWTPayload.js"), exports);
__exportStar(require("./services/jwt/contracts/vcrJWTPayload.js"), exports);
__exportStar(require("./services/jwt/contracts/createVonageTokenParams.js"), exports);
__exportStar(require("./services/jwt/contracts/acl.js"), exports);
__exportStar(require("./services/jwt/contracts/ICreateVonageTokenParams.js"), exports);
__exportStar(require("./services/jwt/contracts/IAcl.js"), exports);
__exportStar(require("./providers/vonageAPI/contracts/invokePayload.js"), exports);
__exportStar(require("./providers/vonageAPI/contracts/IInvokePayload.js"), exports);
__exportStar(require("./providers/voice/enums/voiceAction.js"), exports);
__exportStar(require("./providers/voice/contracts/callEventParams.js"), exports);
__exportStar(require("./providers/voice/contracts/callAnswerCallBack.js"), exports);
__exportStar(require("./providers/voice/contracts/ICallEventParams.js"), exports);
__exportStar(require("./providers/voice/contracts/ICallEventCallBackPayload.js"), exports);
__exportStar(require("./providers/voice/contracts/ICallAnswerCallBack.js"), exports);
__exportStar(require("./providers/voice/contracts/CallEventCallBackPayload.js"), exports);
__exportStar(require("./providers/state/enums/stateOperations.js"), exports);
__exportStar(require("./providers/state/enums/expireOption.js"), exports);
__exportStar(require("./providers/state/contracts/stateCommand.js"), exports);
__exportStar(require("./providers/state/contracts/IStateCommand.js"), exports);
__exportStar(require("./providers/scheduler/enums/schedulerAction.js"), exports);
__exportStar(require("./providers/scheduler/contracts/untilParams.js"), exports);
__exportStar(require("./providers/scheduler/contracts/startAtPayload.js"), exports);
__exportStar(require("./providers/scheduler/contracts/startAtParams.js"), exports);
__exportStar(require("./providers/scheduler/contracts/schedulerPayload.js"), exports);
__exportStar(require("./providers/scheduler/contracts/schedulerIDResponse.js"), exports);
__exportStar(require("./providers/scheduler/contracts/schedulerExecutionInfo.js"), exports);
__exportStar(require("./providers/scheduler/contracts/listAllSchedulersResponse.js"), exports);
__exportStar(require("./providers/scheduler/contracts/listAllPayload.js"), exports);
__exportStar(require("./providers/scheduler/contracts/intervalParams.js"), exports);
__exportStar(require("./providers/scheduler/contracts/getSchedulerResponse.js"), exports);
__exportStar(require("./providers/scheduler/contracts/IUntilParams.js"), exports);
__exportStar(require("./providers/scheduler/contracts/IStartAtPayload.js"), exports);
__exportStar(require("./providers/scheduler/contracts/IStartAtParams.js"), exports);
__exportStar(require("./providers/scheduler/contracts/ISchedulePayload.js"), exports);
__exportStar(require("./providers/scheduler/contracts/IListAllPayload.js"), exports);
__exportStar(require("./providers/scheduler/contracts/IIntervalParams.js"), exports);
__exportStar(require("./providers/queue/contracts/updateQueuePayload.js"), exports);
__exportStar(require("./providers/queue/contracts/updateQueueOptions.js"), exports);
__exportStar(require("./providers/queue/contracts/queueRate.js"), exports);
__exportStar(require("./providers/queue/contracts/queueDetailsResponse.js"), exports);
__exportStar(require("./providers/queue/contracts/createQueuePayload.js"), exports);
__exportStar(require("./providers/queue/contracts/createQueueOptions.js"), exports);
__exportStar(require("./providers/queue/contracts/IUpdateQueuePayload.js"), exports);
__exportStar(require("./providers/queue/contracts/IUpdateQueueOptions.js"), exports);
__exportStar(require("./providers/queue/contracts/IQueueRate.js"), exports);
__exportStar(require("./providers/queue/contracts/IQueueDetailsStats.js"), exports);
__exportStar(require("./providers/queue/contracts/IQueueDetails.js"), exports);
__exportStar(require("./providers/queue/contracts/ICreateQueuePayload.js"), exports);
__exportStar(require("./providers/queue/contracts/ICreateQueueOptions.js"), exports);
__exportStar(require("./providers/messages/contracts/unsubscribeEventsPayload.js"), exports);
__exportStar(require("./providers/messages/contracts/messageContact.js"), exports);
__exportStar(require("./providers/messages/contracts/listenMessagesPayload.js"), exports);
__exportStar(require("./providers/messages/contracts/listenEventsPayload.js"), exports);
__exportStar(require("./providers/messages/contracts/IUnsubscribeEventsPayload.js"), exports);
__exportStar(require("./providers/messages/contracts/IMessageContact.js"), exports);
__exportStar(require("./providers/messages/contracts/IListenMessagesPayload.js"), exports);
__exportStar(require("./providers/messages/contracts/IListenEventsPayload.js"), exports);
__exportStar(require("./providers/messages/enums/messageAction.js"), exports);
__exportStar(require("./providers/logger/enums/sourceType.js"), exports);
__exportStar(require("./providers/logger/enums/logLevel.js"), exports);
__exportStar(require("./providers/events/enums/vcrEventType.js"), exports);
__exportStar(require("./providers/events/enums/vcrEventSourceType.js"), exports);
__exportStar(require("./providers/logger/contracts/logContext.js"), exports);
__exportStar(require("./providers/logger/contracts/logAction.js"), exports);
__exportStar(require("./providers/logger/contracts/ILogContext.js"), exports);
__exportStar(require("./providers/logger/contracts/ILogAction.js"), exports);
__exportStar(require("./providers/assets/enums/fileRetentionPeriod.js"), exports);
__exportStar(require("./providers/assets/enums/assetsAction.js"), exports);
__exportStar(require("./providers/events/contracts/vcrEvent.js"), exports);
__exportStar(require("./providers/events/contracts/sessionCreatedDetails.js"), exports);
__exportStar(require("./providers/events/contracts/IVCREvent.js"), exports);
__exportStar(require("./providers/events/contracts/ISessionCreatedDetails.js"), exports);
__exportStar(require("./providers/assets/contracts/removeAssetPayload.js"), exports);
__exportStar(require("./providers/assets/contracts/listAssetsPayload.js"), exports);
__exportStar(require("./providers/assets/contracts/linkPayload.js"), exports);
__exportStar(require("./providers/assets/contracts/getAssetResponse.js"), exports);
__exportStar(require("./providers/assets/contracts/getAssetPayload.js"), exports);
__exportStar(require("./providers/assets/contracts/directoryPayload.js"), exports);
__exportStar(require("./providers/assets/contracts/assetListResponse.js"), exports);
__exportStar(require("./providers/assets/contracts/assetLinkResponse.js"), exports);
__exportStar(require("./providers/assets/contracts/assetInfo.js"), exports);
__exportStar(require("./providers/conversation/enums/conversationAction.js"), exports);
__exportStar(require("./providers/conversation/contracts/conversationPayloadWithCallback.js"), exports);
__exportStar(require("./providers/conversation/contracts/IConversationEventCallback.js"), exports);
__exportStar(require("./providers/state/enums/fullText/unit.js"), exports);
__exportStar(require("./providers/state/enums/fullText/tags.js"), exports);
__exportStar(require("./providers/state/enums/fullText/summarizeOption.js"), exports);
__exportStar(require("./providers/state/enums/fullText/sortOrder.js"), exports);
__exportStar(require("./providers/state/enums/fullText/sortByOption.js"), exports);
__exportStar(require("./providers/state/enums/fullText/searchOptions.js"), exports);
__exportStar(require("./providers/state/enums/fullText/returnOption.js"), exports);
__exportStar(require("./providers/state/enums/fullText/paramOption.js"), exports);
__exportStar(require("./providers/state/enums/fullText/limitOption.js"), exports);
__exportStar(require("./providers/state/enums/fullText/inKeysOptions.js"), exports);
__exportStar(require("./providers/state/enums/fullText/inFieldsOption.js"), exports);
__exportStar(require("./providers/state/enums/fullText/highlightOption.js"), exports);
__exportStar(require("./providers/state/enums/fullText/geoFilterOption.js"), exports);
__exportStar(require("./providers/state/enums/fullText/filterOption.js"), exports);
__exportStar(require("./providers/state/enums/fullText/fieldProperty.js"), exports);
__exportStar(require("./providers/state/enums/fullText/IndexType.js"), exports);
__exportStar(require("./providers/state/enums/fullText/ITags.js"), exports);
__exportStar(require("./providers/state/enums/fullText/ISummarizeOption.js"), exports);
__exportStar(require("./providers/state/enums/fullText/ISortByOption.js"), exports);
__exportStar(require("./providers/state/enums/fullText/ISearchOptions.js"), exports);
__exportStar(require("./providers/state/enums/fullText/IReturnOption.js"), exports);
__exportStar(require("./providers/state/enums/fullText/IParamOption.js"), exports);
__exportStar(require("./providers/state/enums/fullText/ILimitOption.js"), exports);
__exportStar(require("./providers/state/enums/fullText/IInKeysOption.js"), exports);
__exportStar(require("./providers/state/enums/fullText/IInFieldsOption.js"), exports);
__exportStar(require("./providers/state/enums/fullText/IHighlightOption.js"), exports);
__exportStar(require("./providers/state/enums/fullText/IGeoFilterOption.js"), exports);
__exportStar(require("./providers/state/enums/fullText/IFilterOption.js"), exports);
__exportStar(require("./providers/state/enums/fullText/IFieldProperty.js"), exports);
__exportStar(require("./providers/state/enums/fullText/FieldType.js"), exports);
__exportStar(require("./providers/state/contracts/fullText/stopwordsOptions.js"), exports);
__exportStar(require("./providers/state/contracts/fullText/schemaField.js"), exports);
__exportStar(require("./providers/state/contracts/fullText/prefixOptions.js"), exports);
__exportStar(require("./providers/state/contracts/fullText/createIndexOptions.js"), exports);
__exportStar(require("./providers/state/contracts/fullText/IStopwordsOptions.js"), exports);
__exportStar(require("./providers/state/contracts/fullText/ISchemaField.js"), exports);
__exportStar(require("./providers/state/contracts/fullText/IPrefixOptions.js"), exports);
__exportStar(require("./providers/state/contracts/fullText/ICreateIndexOptions.js"), exports);
