// neru-alpha/dist/cjs/providers/voice/csEvents.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CSEvents = void 0;
var CSEvents;
(function (CSEvents) {
    CSEvents["AppKnocking"] = "app:knocking";
    CSEvents["AudioSay"] = "audio:say";
    CSEvents["AudioSayStop"] = "audio:say:stop";
    CSEvents["AudioSayDone"] = "audio:say:done";
    CSEvents["AudioPlay"] = "audio:play";
    CSEvents["AudioPlayStop"] = "audio:play:stop";
    CSEvents["AudioPlayDone"] = "audio:play:done";
    CSEvents["AudioDTMF"] = "audio:dtmf";
    CSEvents["EarmuffOn"] = "audio:earmuff:on";
    CSEvents["EarmuffOff"] = "audio:earmuff:off";
    CSEvents["MuteOn"] = "audio:mute:on";
    CSEvents["MuteOff"] = "audio:mute:off";
    CSEvents["LegStatusUpdate"] = "leg:status:update";
    CSEvents["MemberJoined"] = "member:joined";
    CSEvents["MemberInvited"] = "member:invited";
    CSEvents["MemberLeft"] = "member:left";
    CSEvents["ConversationCreated"] = "conversation:created";
})(CSEvents = exports.CSEvents || (exports.CSEvents = {}));
