// @vonage/vcr-sdk/dist/esm/providers/queue/queue.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { CreateQueuePayload } from "./contracts/createQueuePayload.js";
import { QueueRate } from "./contracts/queueRate.js";
import { RequestParams } from "../../request/contracts/requestParams.js";
import { REQUEST_VERB } from "../../request/enums/requestVerb.js";
import { UpdateQueuePayload } from "./contracts/updateQueuePayload.js";
var Queue = /** @class */ (function () {
    function Queue(session) {
        this.provider = "queue-service";
        this.session = session;
        this.bridge = session.bridge;
        this.config = session.config;
    }
    Queue.prototype.createQueue = function (name, callback, options) {
        return __awaiter(this, void 0, void 0, function () {
            var payload, requestParams;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        payload = new CreateQueuePayload();
                        payload.name = name;
                        payload.callback = this.session.wrapCallback(callback, []);
                        payload.active = options.active;
                        payload.rate = new QueueRate();
                        payload.rate.maxInflight = options.maxInflight;
                        payload.rate.msgPerSecond = options.msgPerSecond;
                        requestParams = new RequestParams();
                        requestParams.method = REQUEST_VERB.POST;
                        requestParams.data = payload;
                        requestParams.url = this.config.getExecutionUrl(this.provider, "queue");
                        requestParams.headers = this.session.constructRequestHeaders();
                        return [4 /*yield*/, this.session.request(requestParams)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    Queue.prototype.updateQueue = function (queueName, options) {
        return __awaiter(this, void 0, void 0, function () {
            var payload, requestParams;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        payload = new UpdateQueuePayload();
                        payload.rate = new QueueRate();
                        payload.rate.maxInflight = options.maxInflight;
                        payload.rate.msgPerSecond = options.msgPerSecond;
                        requestParams = new RequestParams();
                        requestParams.method = REQUEST_VERB.POST;
                        requestParams.data = payload;
                        requestParams.url = this.config.getExecutionUrl(this.provider, "queue/".concat(queueName));
                        requestParams.headers = this.session.constructRequestHeaders();
                        return [4 /*yield*/, this.session.request(requestParams)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    Queue.prototype.list = function () {
        return __awaiter(this, void 0, void 0, function () {
            var requestParams;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        requestParams = new RequestParams();
                        requestParams.method = REQUEST_VERB.GET;
                        requestParams.data = null;
                        requestParams.url = this.config.getExecutionUrl(this.provider, "queue");
                        requestParams.headers = this.session.constructRequestHeaders();
                        return [4 /*yield*/, this.session.request(requestParams)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    Queue.prototype.getQueueDetails = function (name) {
        return __awaiter(this, void 0, void 0, function () {
            var requestParams;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        requestParams = new RequestParams();
                        requestParams.method = REQUEST_VERB.GET;
                        requestParams.data = null;
                        requestParams.url = this.config.getExecutionUrl(this.provider, "queue/".concat(name));
                        requestParams.headers = this.session.constructRequestHeaders();
                        return [4 /*yield*/, this.session.request(requestParams)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    Queue.prototype.deleteQueue = function (name) {
        return __awaiter(this, void 0, void 0, function () {
            var requestParams;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        requestParams = new RequestParams();
                        requestParams.method = REQUEST_VERB.DEL;
                        requestParams.data = null;
                        requestParams.url = this.config.getExecutionUrl(this.provider, "queue/".concat(name));
                        requestParams.headers = this.session.constructRequestHeaders();
                        return [4 /*yield*/, this.session.request(requestParams)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    Queue.prototype.pauseQueue = function (name) {
        return __awaiter(this, void 0, void 0, function () {
            var requestParams;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        requestParams = new RequestParams();
                        requestParams.method = REQUEST_VERB.PUT;
                        requestParams.data = null;
                        requestParams.url = this.config.getExecutionUrl(this.provider, "queue/".concat(name, "/pause"));
                        requestParams.headers = this.session.constructRequestHeaders();
                        return [4 /*yield*/, this.session.request(requestParams)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    Queue.prototype.resumeQueue = function (name) {
        return __awaiter(this, void 0, void 0, function () {
            var requestParams;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        requestParams = new RequestParams();
                        requestParams.method = REQUEST_VERB.PUT;
                        requestParams.data = null;
                        requestParams.url = this.config.getExecutionUrl(this.provider, "queue/".concat(name, "/resume"));
                        requestParams.headers = this.session.constructRequestHeaders();
                        return [4 /*yield*/, this.session.request(requestParams)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    Queue.prototype.enqueue = function (name, data) {
        return __awaiter(this, void 0, void 0, function () {
            var requestParams;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        requestParams = new RequestParams();
                        requestParams.method = REQUEST_VERB.POST;
                        requestParams.data = data;
                        requestParams.url = this.config.getExecutionUrl(this.provider, "queue/".concat(name, "/enqueue"));
                        requestParams.headers = this.session.constructRequestHeaders();
                        return [4 /*yield*/, this.session.request(requestParams)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    Queue.prototype.enqueueSingle = function (name, data) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.enqueue(name, [data])];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    Queue.prototype.deadLetterList = function (name) {
        return __awaiter(this, void 0, void 0, function () {
            var requestParams;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        requestParams = new RequestParams();
                        requestParams.method = REQUEST_VERB.GET;
                        requestParams.data = null;
                        requestParams.url = this.config.getExecutionUrl(this.provider, "queue/".concat(name, "/deadletter"));
                        requestParams.headers = this.session.constructRequestHeaders();
                        return [4 /*yield*/, this.session.request(requestParams)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    Queue.prototype.deadLetterDequeue = function (name, count) {
        if (count === void 0) { count = 1; }
        return __awaiter(this, void 0, void 0, function () {
            var requestParams;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        requestParams = new RequestParams();
                        requestParams.method = REQUEST_VERB.POST;
                        requestParams.data = null;
                        requestParams.url = this.config.getExecutionUrl(this.provider, "queue/".concat(name, "/deadletter/pop"), {
                            count: this.bridge.jsonStringify(count),
                        });
                        requestParams.headers = this.session.constructRequestHeaders();
                        return [4 /*yield*/, this.session.request(requestParams)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    return Queue;
}());
export { Queue };
